proof-of-spacetime.md

```diff
@@ -0,0 +1,200 @@
+## Proof-of-Spacetime
+
+This document descibes Rational-PoSt, the Proof of Spacetime used in Filecoin.
+
+
+## Rational PoSt
+
+### Definitions
+
+- **PoSt Proving Period**: The time interval in which a PoSt has to be submitted.
+- **POST_CHALLENGE_TIME**: The time offset at which the actual work of generating the PoSt should be started. This is some delta befre the end of the `Proving Period`, and as such less then a single `Proving Period`.
+
+- start = height x
+- challenge time = height x + y
+- on post submission: verify that challenge comes from block(x + y)
+
+### Execution Flow
+
+```
+    ■──────────────────────── Proving Period ─────────────────────────■
+
+
+    ▲───────────────────────────────────────────────────▲─────────────▲
+    │                                                   │             │
+    │                                                   │             │
+    │                                                   │             │
+┌───────┐                                     ┌──────────────────┐┌───────┐
+│ Start │                                     │ Generation Start ││  End  │
+└───────┘                                     └──────────────────┘└───────┘
+                                                        ▲
+
+                                                        │
+                                                                       ┌ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+                                                        └ ─ ─ ─ ─ ─ ─ ─ Randomness Input
+                                                                       └ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+TODO: Add post submission to the diagram.
+
+
+
+
+### High Level API
+
+#### Generation
+
+```go
+func GeneratePoSt(sectorSize BytesAmount, sectors []commR) (PoStProof, FaultSet) {
+    // Generate the Merkle Inclusion Proofs + Faults
+
+    inclusionProofs := []
+    faults := NewFaultSet()
+
+    for n in 0..POST_CHALLENGES_COUNT {
+        attempt := 0
+        'inner: for {
+            challenge := DerivePoStChallenge(seed, n, faults, attempt)
+            sector := challenge % sectorSize
+            // check if we landed in a previously marked faulty sector
+            if !faults.Contains(sector) {
+                attempt += 1
+                continue
+            }
+
+            challenge_value = challenge / sectorSize
+            inclusionProof, isFault := GenerateMerkleInclusionProof(sector, challenge_value)
+            if isFault {
+                // faulty sector, generate a new challenge
+                faults.Add(sector)
+                attempt += 1
+            } else {
+                // no fault, move on to the next challenge
+                inclusionProofs[n] = inclusionProof
+                break 'inner
+            }
+        }
+    }
```

Comment:
I am actually not too sure if I buy this behavior: if you announce faults only when challenged on them, then you might have a lot more of unreported faults, and this is bad, since if the challenges are ~200, then miners would only report at most 200 of them.

Hence, we:
- EITHER challenge every sector at least once (seems pretty expensive since we want to support 16k sectors per post (in case we do so, we must use the CommA trick, instead of passing one CommR per sector)
- OR miners must announce their faults before the POST_CHALLENGE_TIME, in this way they are required to report them all.

Reply:
Faults need to be announced as per the last discussions.